File: clase_15/code/t1_lock.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <sys/time.h>

/*
1. To compile:

gcc -o t1_lock t1_lock.c -lpthread -Wall

2. To run

./t1_lock <MAX_VALUE>
*/

int max;
volatile int counter = 0; // shared global variable

pthread_mutex_t lock; // lock variable (Mutex)

void *mythread(void *arg) {
    char *letter = arg;
    int i; // stack (private per thread)     
    for (i = 0; i < max; i++) {
        pthread_mutex_lock(&lock);      // -------------------------------------- Acquire the lock (adquire(L)) ----------------------------------------------
        counter = counter + 1;          // critical section: only one
        pthread_mutex_unlock(&lock);    // -------------------------------------- Release the lock (release(L)) ----------------------------------------------
    }
    printf("%s: done\n", letter);
    return NULL;
}
                                                                             
int main(int argc, char *argv[]) {                    
    if (argc != 2) {
        fprintf(stderr, "usage: main-first <loopcount>\n");
        exit(1);
    }
    max = atoi(argv[1]);

    pthread_t p1, p2;
    int rc = pthread_mutex_init(&lock, NULL); // Initialize the lock
    if (rc != 0) {
        fprintf(stderr, "main: mutex init failed\n");
        exit(1);
    }
    printf("main: begin [counter = %d]\n", counter);
    pthread_create(&p1, NULL, mythread, "A"); 
    pthread_create(&p2, NULL, mythread, "B");
    // join waits for the threads to finish
    pthread_join(p1, NULL); 
    pthread_join(p2, NULL); 
    printf("main: done\n [counter: %d]\n [should: %d]\n", 
	   counter, max*2);
    return 0;
}

File: clase_15/code/t1_lock.py
import threading
import sys

# Necesario usar un objeto para que sea mutable y compartido entre hilos
class Counter:
    def __init__(self):
        self.value = 0

counter = Counter()
lock = threading.Lock() # Crear un lock global
max = 0

def mythread(arg):
  global counter
  # stack (private per thread)   
  print(f"{arg}: begin")
  for i in range(max):
    with lock:               # Context manager: adquiere y libera el lock automaticamente
       # --- INICIO DE LA SECCIÓN CRÍTICA ---
       counter.value += 1
       # --- FIN DE LA SECCIÓN CRÍTICA ---
    # Simular a las llamadas pthread_mutex_lock y pthread_mutex_unlock
    # lock.acquire()          # pthread_mutex_lock()    
    # counter.value += 1
    # lock.release()          # pthread_mutex_unlock()
    
  print(f"{arg}: done")

if __name__ == "__main__":
  if len(sys.argv) != 2:
    print("usage: main <loopcount>")
  else:
    max = int(sys.argv[1])
    print("main: begin")
    print(f"main: begin [counter = {counter.value}]")
  
    # Creacion de los hilos
    p1 = threading.Thread(target=mythread, args=('A',))
    p2 = threading.Thread(target=mythread, args=('B',))
    # Se inicia la ejecucion de los hilos
    p1.start()
    p2.start()
    # El hilo padre espera que los hilos hijos culminen
    p1.join()
    p2.join()
    # Mensaje que se imprime despues de que el padre reinicia su ejecucion
    print(f"end: begin [counter = {counter.value}]")
    print(f" [counter: {counter.value}]\n [should: {max*2}]\n")